### PR TITLE
c11_atomics: Fix to iteratively reduce workgroup size  (#939)

### PR DESCRIPTION
**Pull request by lakshmih**
The generator function for atomic_fence uses the current workgroup
size to determine the number of non atomic variables per thread
in the kernel. The kernel should hence be regenerated when the
workgroup size changes.

However regenerating the kernel can itself change the workgroup
size. This change introduces an iterative loop that reduces the
workgroup sizes by 2 each time re-generating the kernel until
we find one that works (or exit at groupsize == 1)

Change-Id: Ic32fe967e32de6643e01c6775f4bddbcad0a299a

### CODE REVIEW

**Comment by lakshmih** [comment]
@kpet @bashbaug  Please review. This is a fix that was originally merged to the cl20_trunk branch on gitlab (links to old bugzilla issue and fix on the issue description) and not propagated to the 2.1 and future branches.
Please add anyone else that you think needs to be on the review.

File: test_conformance/c11_atomics/common.h [hunk 1]
```diff
@@ -948,51 +948,76 @@ int CBasicTest<HostAtomicType, HostDataType>::ExecuteSingleTest(cl_device_id dev
...
+      {
+          // Re-generate the kernel code with the current group size
+          if (kernel) clReleaseKernel(kernel);
+          if (program) clReleaseProgram(program);
+          programSource = PragmaHeader(deviceID) + ProgramHeader(numDestItems)
+              + FunctionCode() + KernelCode(numDestItems);
+          programLine = programSource.c_str();
+          if (create_single_kernel_helper_with_build_options(
+                  context, &program, &kernel, 1, &programLine,
+                  "test_atomic_kernel", gOldAPI ? "" : "-cl-std=CL2.0"))
```
Comment (Jeremy-Kemp) [inline test_conformance/c11_atomics/common.h:968]: Was it intentional to go back to specifying `-cl-std=CL2.0`?

Comment (lakshmih) [reply test_conformance/c11_atomics/common.h:968]: oops no, that was a typo. Thanks for catching it, I'll update.

Comment (lakshmih) [reply test_conformance/c11_atomics/common.h]: @Jeremy-Kemp can you review again? Thanks.

**Comment by lakshmih** [comment]
Copying notes from issue description for reviewers:

Original issue:
https://www.khronos.org/members/login/bugzilla/show_bug.cgi?id=15676

Original fix merged to cl20_trunk:
https://gitlab.khronos.org/opencl/conformance-tests/-/commit/b71bb46fe0dd4aee9f5d74571a6b6982de24cf96

I've re-summarized it below.

The generator function for atomic_fence uses the current workgroup size to determine the number of non atomic variables per thread. However, when the workgroup size is modified (after the kernel is generated) we don't re-generate the kernel, leading to mismatches during verification.

Following is the sequence in file
https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/c11_atomics/common.h
1)
CurrentGroupSize is set to the maximum workgroup size allowed by the device
// set up work sizes based on device capabilities and test configuration
error = clGetDeviceInfo(deviceID, CL_DEVICE_MAX_WORK_GROUP_SIZE, sizeof(groupSize), &groupSize, NULL);
test_error(error, "Unable to obtain max work group size for device");
CurrentGroupSize((cl_uint)groupSize);

2)
The CL kernel is generated. The generator function for the kernel uses the workgroup size in determining NumNonAtomicVariablesPerThread(). See calculation of NumNonAtomicVariablesPerThread in test_atomics.cpp

3)
Workgroup size is modified depending on the max permitted for the kernel

// tune up work sizes based on kernel info
error = clGetKernelWorkGroupInfo(kernel, deviceID, CL_KERNEL_WORK_GROUP_SIZE, sizeof(groupSize), &groupSize, NULL);
test_error(error, "Unable to obtain max work group size for device and kernel combo");

If the workgroup size changes in step 3 above, the kernel needs to be re-generated. If not, the verify function uses the modified (kernel_max) workgroup size to calculate the reference values whereas the kernel was generated using the original (device_max) workgroup size.

**Review by Jeremy-Kemp — APPROVED** [review]
Thanks for the fix @lakshmih .

LGTM, especially as we had this in GitLab. I've sanity checked this on our implementation as well.

**Review by bashbaug — APPROVED** [review]
LGTM, especially if these changes were already on GitLab.

**Comment by bashbaug** [comment]
Adding to the OpenCL 3.0 board for more visibility, though this is not a "must fix for OpenCL 3.0" issue.

**Comment by lakshmih** [comment]
Thanks for verifying @Jeremy-Kemp  
@kpet  can you merge this? I'm still not sure about the protocol for merging stuff.

**Pull request merged**